### PR TITLE
Create Relationship For Character and Vote

### PR DESCRIPTION
**Pull request by stefanbesovic**
Updated Character and Vote entities with necessary fields, and implemented logic for Adding and Removing in Services and Controllers.

### CODE REVIEW

File: src/main/java/com/axiomq/starwars/entities/Character.java [hunk 1]
```diff
@@ -37,6 +37,15 @@
     private Planet planet;
     private Integer votersCount;
 
+    @OneToMany(mappedBy = "character", orphanRemoval = true)
+    private Set<Vote> votes = new HashSet<>();
+
+    @ElementCollection
+    private Set<String> usersEmail;
```
Comment (branimir-vujicic) [inline src/main/java/com/axiomq/starwars/entities/Character.java:44]: add relation (many to many ? ) to user

File: src/main/java/com/axiomq/starwars/services/impl/CharacterServiceImpl.java [hunk 2]
```diff
@@ -59,4 +59,20 @@ public void populateCharacters() {
         ResourceDatabasePopulator databasePopulator = new ResourceDatabasePopulator(resource);
         databasePopulator.execute(dataSource);
     }
+
+    @Override
+    public void addCharacterVotersCount(Long characterId, Principal principal) {
+        Character character = getCharacterById(characterId);
+        character.addEmail(principal.getName());
+        character.setVotersCount(character.getUsersEmail().size());
```
Comment (branimir-vujicic) [inline src/main/java/com/axiomq/starwars/services/impl/CharacterServiceImpl.java:67]: user repository method count instead

File: pom.xml [hunk 3]
```diff
@@ -93,6 +93,10 @@
 			<version>0.11.2</version>
 			<scope>runtime</scope>
 		</dependency>
+		<dependency>
+			<groupId>org.springframework.boot</groupId>
+			<artifactId>spring-boot-starter-webflux</artifactId>
+		</dependency>
```
Comment (branimir-vujicic) [inline pom.xml:99]: If you using rest template, this dependency is obsolete

File: src/main/java/com/axiomq/starwars/repositories/VoteRepository.java [hunk 4]
```diff
@@ -2,6 +2,12 @@
...
 import com.axiomq.starwars.entities.Vote;
 import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.data.jpa.repository.Query;
 
 public interface VoteRepository extends JpaRepository<Vote, Long> {
+
+    @Query(
+            value = "SELECT COUNT(DISTINCT v.user_id) FROM vote v WHERE v.character_id = ?1",
+            nativeQuery = true)
+    Integer countDistinctUsers(Long characterId);
```
Comment (branimir-vujicic) [inline src/main/java/com/axiomq/starwars/repositories/VoteRepository.java:12]: you could use `long countDistinctByUser(User user);` instead native query

File: src/main/java/com/axiomq/starwars/services/impl/CharacterImportServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+
+        return extractCharacters(response);
+    }
+
+    public Set<Character> extractCharacters(CharacterGet response) {
+        Set<Character> real = new HashSet<>();
+
+        List<CharacterResponse> characters = new ArrayList<>(response.getResults());
+        characters.forEach(characterResponse ->
+                real.add(convertCharacter.toCharacter(characterResponse)));
```
Comment (branimir-vujicic) [inline src/main/java/com/axiomq/starwars/services/impl/CharacterImportServiceImpl.java:40]: it's more readable if you use streams
`Set<Character> real = response.getResults.stream()
    .map(characterResponse -> convertCharacter.toCharacter(characterResponse))
    .collect(Collectors.toSet());`

File: src/main/java/com/axiomq/starwars/repositories/VoteRepository.java (hunk 4 above)
Comment (stefanbesovic) [reply src/main/java/com/axiomq/starwars/repositories/VoteRepository.java:12]: how would i know if the returned number of users is only for my character's vote, not from the others ? that's why i need the 
'v.character_id = ?1', right ?

Comment (branimir-vujicic) [reply src/main/java/com/axiomq/starwars/repositories/VoteRepository.java:12]: long countDistinctUserByCharacter(Character character);

**Pull request merged**